channel2/wss_listener.go

```diff
@@ -0,0 +1,280 @@
+	return &wssListener{
+		identity:       identity,
+		endpoint:       endpoint,
+		close:          make(chan struct{}),
+		created:        make(chan Underlay),
+		incoming:       make(chan *WssConnection, connectOptions.MaxQueuedConnects),
+		connectOptions: connectOptions,
+		serverCert:     serverCert,
+		key:            key,
+		tcfg:           tcfg,
+	}
+}
+
+type WssConnection struct {
+	detail    *transport.ConnectionDetail
+	websocket *websocket.Conn
+	socket    *tls.Conn
+}
+
+func (c *WssConnection) Detail() *transport.ConnectionDetail {
+	return c.detail
+}
+
+func (c *WssConnection) PeerCertificates() []*x509.Certificate {
+	return c.socket.ConnectionState().PeerCertificates
+}
+
+func (c *WssConnection) Reader() io.Reader {
+	return c.socket
+}
+
+func (c *WssConnection) Writer() io.Writer {
+	return c.socket
+}
+
+func (c *WssConnection) Websocket() *websocket.Conn {
+	return c.websocket
+}
+
+func (c *WssConnection) Conn() net.Conn {
+	return c.socket
+}
+
+func (c *WssConnection) SetReadTimeout(t time.Duration) error {
+	return c.socket.SetReadDeadline(time.Now().Add(t))
+}
+
+func (c *WssConnection) SetWriteTimeout(t time.Duration) error {
+	return c.socket.SetWriteDeadline(time.Now().Add(t))
+}
+
+func (c *WssConnection) Close() error {
+	return c.socket.Close()
+}
+
+var upgrader = websocket.Upgrader{} // use default options
+
+/**
+ *	Accept incoming HTTP connection, and upgrade it to a websocket suitable for comms between browZer and Ziti edge router
+ */
+func (listener *wssListener) handleWebsocket(w http.ResponseWriter, r *http.Request) {
+	log := pfxlog.ContextLogger(listener.endpoint.String())
+	log.Info("entered")
+
+	upgrader.CheckOrigin = func(r *http.Request) bool { return true } // Allow all origins
+
+	c, err := upgrader.Upgrade(w, r, nil) // upgrade from HTTP to binary socket
+
+	if err != nil {
+		log.WithField("err", err).Error("websocket upgrade failed. Failure not recoverable. Exiting listen loop")
+	} else {
+
+		socket := c.UnderlyingConn() // Obtain the socket underneath the websocket
+
+		connection := &WssConnection{
+			detail: &transport.ConnectionDetail{
+				Address: "wss:" + socket.RemoteAddr().String(),
+				InBound: true,
+				Name:    "wss",
+			},
+			websocket: c,
+			socket:    socket.(*tls.Conn),
+		}
+
+		listener.incoming <- connection // pass the socket to the goroutine that will validate the HELLO handshake
+	}
+}
+
+/**
+ *	Simply start the goroutines that will process incoming websocket connections
+ */
+func (listener *wssListener) Listen(handlers ...ConnectionHandler) error {
+	go listener.wsslistener()
+	go listener.binaryListener()
+	return nil
+}
+
+/**
+ *	The TLS-based listener that accepts incoming HTTP connections that we need to upgrade to websocket connections
+ */
+func (listener *wssListener) wsslistener() {
+	log := pfxlog.ContextLogger(listener.endpoint.String())
+	log.Info("starting HTTP (websocket) server")
+
+	router := mux.NewRouter()
```

Comment:
The edge no longer uses `gorilla.mux`. It looks a bit overkill to bring it in here just for this one endpoint.

Reply:
obviated by new wss transport... so removed